stdlib/LinearAlgebra/src/matmul.jl

```diff
@@ -1081,3 +1081,129 @@ function matmul3x3!(C::AbstractMatrix, tA, tB, A::AbstractMatrix, B::AbstractMat
     end # inbounds
     C
 end
+
+const RealOrComplex = Union{Real,Complex}
+
+# Three-argument *
+"""
+    *(A, B::AbstractMatrix, C)
+
+Most 3- and 4-argument `*` calls containing matrices or vectors are done in an
+efficient way, rather than the left-to-right default of `*(x,y,z,...)`.
+
+This can mean performing `B*C` first if `C::AbstractVector`, using five-argument
+`mul!` to fuse the scalar `A::Number` with the matrix multiplication `B*C`,
+or examining `size.((A,B,C,D))` to choose which to multiply first.
+
+!!! compat "Julia 1.6"
+    These optimisations require least Julia 1.6.
+"""
+*(A::AbstractMatrix, B::AbstractMatrix, x::AbstractVector) = A * (B*x)
+
+*(A::AbstractMatrix, x::AbstractVector, γ::Number) = mat_vec_scalar(A,x,γ)
+*(A::AbstractMatrix, B::AbstractMatrix, γ::Number) = mat_mat_scalar(A,B,γ)
+*(α::RealOrComplex, B::AbstractMatrix{<:RealOrComplex}, C::AbstractVector{<:RealOrComplex}) =
+    mat_vec_scalar(B,C,α)
+*(α::RealOrComplex, B::AbstractMatrix{<:RealOrComplex}, C::AbstractMatrix{<:RealOrComplex}) =
+    mat_mat_scalar(B,C,α)
+
+*(α::Number, u::AbstractVector, tv::AdjOrTransAbsVec) = broadcast(*, α, u, tv)
+*(u::AbstractVector, tv::AdjOrTransAbsVec, γ::Number) = broadcast(*, u, tv, γ)
+*(u::AbstractVector, tv::AdjOrTransAbsVec, C::AbstractMatrix) = u * (tv*C)
+
+*(A::AbstractMatrix, B::AbstractMatrix, C::AbstractMatrix) = _tri_matmul(A,B,C)
+*(tv::AdjOrTransAbsVec, B::AbstractMatrix, C::AbstractMatrix) = (tv*B) * C
+
+function _tri_matmul(A,B,C,δ=nothing)
+    n,m = size(A)
+    # m,k == size(B)
+    k,l = size(C)
+    costAB_C = n*m*k + n*k*l
+    costA_BC = m*k*l + n*m*l
+    if costA_BC < costAB_C
+        isnothing(δ) ? A * (B*C) : A * mat_mat_scalar(B,C,δ)
+    else
+        isnothing(δ) ? (A*B) * C : mat_mat_scalar(A*B, C, δ)
+    end
+end
+
+# Fast path for two arrays * one scalar is opt-in, via mat_vec_scalar and mat_mat_scalar.
+
+mat_vec_scalar(A, x, γ) = A * (x .* γ)  # fallback
+mat_vec_scalar(A::StridedMaybeAdjOrTransMat, x::StridedVector, γ) = _mat_vec_scalar(A, x, γ)
+mat_vec_scalar(A::AdjOrTransAbsVec, x::StridedVector, γ) = (A * x) * γ
+
+function _mat_vec_scalar(A, x, γ)
+    T = promote_type(eltype(A), eltype(x), typeof(γ))
+    C = similar(A, T, axes(A,1))
+    mul!(C, A, x, γ, false)
+end
+
+mat_mat_scalar(A, B, γ) = (A*B) .* γ # fallback
+mat_mat_scalar(A::StridedMaybeAdjOrTransMat, B::StridedMaybeAdjOrTransMat, γ) =
+    _mat_mat_scalar(A, B, γ)
+
+function _mat_mat_scalar(A, B, γ)
+    T = promote_type(eltype(A), eltype(B), typeof(γ))
+    C = similar(A, T, axes(A,1), axes(B,2))
+    mul!(C, A, B, γ, false)
+end
+
+mat_mat_scalar(A::AdjointAbsVec, B, γ) = (γ' .* (A * B)')' # preserving order, adjoint reverses
+mat_mat_scalar(A::AdjointAbsVec, B::StridedMaybeAdjOrTransMat, γ::Union{Real,Complex}) =
```

Comment:
```suggestion
mat_mat_scalar(A::AdjointAbsVec, B::StridedMaybeAdjOrTransMat, γ::RealOrComplex) =
```
Doesn't this require `A` and `B` to be of `RealOrComplex` eltype?

Reply:
I think LinearAlgebra assumes that `(A * B)' == B' * A'` holds in general, both for matrices and their elements. Certainly this should be true for matrices-of-matrices, I have not checked but I would guess that sensible quaternions would also obey this.

Reply:
Sure, but taking the adjoint reverse the order of all factors, including gamma, which does not change position here. I double checked, and found this:
```julia
julia> using Quaternions

julia> q1 = quat(rand(4))
Quaternion{Float64}(0.0, 0.7878876634570733, 0.5833693013402739, 0.41287390303097493, false)

julia> q2 = quat(rand(4))
Quaternion{Float64}(0.0, 0.16366198479946625, 0.29162644956662964, 0.7409506995819128, false)

julia> z = quat(rand(2)..., 0, 0)
Quaternion{Float64}(0.934172670739851, 0.47087426101844043, 0.0, 0.0, false)

julia> (q1*q2*z)' ≈ q2'*q1'*z'
false
```

Reply:
Oh right, sorry, I agree with you now. Will constrain the eltypes here.